voice/src/main/kotlin/VoiceConnection.kt

```diff
@@ -0,0 +1,146 @@
+ * @param guildId the id of the guild that the bot is connecting to.
+ * @param sessionId the id of the current voice session, given by Discord.
+ */
+data class VoiceConnectionData(
+    val selfId: Snowflake,
+    val guildId: Snowflake,
+    val sessionId: String
+)
+
+private val voiceConnectionLogger = KotlinLogging.logger { }
+
+/**
+ * A [VoiceConnection] is an adapter that forms the concept of a voice connection, or a connection between you and Discord voice servers.
+ *
+ * @param gateway the [Gateway] that handles events for the guild this [VoiceConnection] represents.
+ * @param voiceGateway the underlying [VoiceGateway] for this voice connection.
+ * @param data the data representing this [VoiceConnection].
+ * @param voiceGatewayConfiguration the configuration used for [voiceGateway].
+ * @param audioProvider a [AudioProvider] that will provide [AudioFrame] when required.
+ * @param frameInterceptorFactory a factory for [FrameInterceptor]s that is used whenever audio is ready to be sent. See [FrameInterceptor] and [DefaultFrameInterceptor].
+ * @param voiceDispatcher the dispatcher used for this voice connection.
+ */
+@KordVoice
+class VoiceConnection(
+    val gateway: Gateway,
+    val voiceGateway: VoiceGateway,
+    val data: VoiceConnectionData,
+    internal var voiceGatewayConfiguration: VoiceGatewayConfiguration,
+    var audioProvider: AudioProvider,
+    var frameInterceptorFactory: (FrameInterceptorContext) -> FrameInterceptor,
+    voiceDispatcher: CoroutineDispatcher
+) : CoroutineScope {
+    override val coroutineContext: CoroutineContext =
+        SupervisorJob() + voiceDispatcher + CoroutineName("Voice Connection for Guild ${data.guildId.value}")
+
+    private val audioFramePoller = AudioFramePoller(TimeSource.Monotonic, voiceDispatcher)
+
+    init {
+        // handle voice state/server updates (e.g., a move, disconnect, voice server change, etc.)
+        VoiceUpdateEventHandler(this, gateway.events)
+
+        // handle the lifecycle of the udp connection
+        UdpLifeCycleHandler(
+            voiceGateway.events,
+            voiceGateway::send,
+            { audioFramePoller.start(it.withConnection(this)) },
+            voiceDispatcher
+        )
+    }
+
+    /**
+     * Logs into the voice gateway, and begins the process of an audio-ready voice session.
+     */
+    fun connect() {
+        launch {
+            voiceGateway.start(voiceGatewayConfiguration)
+        }
+    }
+
+    /**
+     * Disconnects from the voice gateway, does not change the voice state.
+     */
+    suspend fun disconnect() {
+        voiceGateway.stop()
+    }
+
+    /**
+     * Disconnects from Discord voice servers, and leaves the voice channel.
+     */
+    suspend fun leave() {
+        disconnect()
+
+        gateway.send(
+            UpdateVoiceStatus(
+                guildId = data.guildId,
+                channelId = null,
+                selfMute = false,
+                selfDeaf = false
```

Comment:
are those mandatory?

Reply:
According to Discord docs, they are not nullable, so yes.